torch/export/dynamic_shapes.py

```diff
@@ -1191,3 +1193,280 @@ def apply_fixes(path, dim, dummy):
+        return out
+
+    def _track_dim_from_dims(val: Union[None, int, _DimHint, _Dim]) -> Union[None, int, str]:
+        """
+        Tracks dims, ranges, derived dims from the standardized dynamic_shapes spec.
+        """
+        if val is None or isinstance(val, int):  # non-tensor input or static
+            return val
+        if isinstance(val, _DimHint):  # store enum as string
+            return val.__class__.__name__ + '.' + val.name
+
+        assert isinstance(val, _Dim)
+
+        # track root dim
+        root = val.root if isinstance(val, _DerivedDim) else val  # type: ignore[attr-defined]
+        if root.__name__ not in dims:
+            dims[root.__name__] = {
+                "min": root.min,
+                "max": root.max,
+                "derived": set(),
+            }
+
+        # track derived dims
+        if isinstance(val, _DerivedDim):
+            dims[root.__name__]["derived"].add(val.__name__)
+
+        return val.__name__
+
+    if dynamic_shapes is None:
+        return {"dynamic_shapes": None, "dims": {}}
+
+    # convert to tuple of specs, for each arg/kwarg
+    kwargs = kwargs or {}
+    if isinstance(dynamic_shapes, dict):
+        dynamic_shapes = dynamic_shapes.values()
+    dynamic_shapes = tuple(dynamic_shapes)
+    combined_args = tuple(args) + tuple(kwargs.values())
+
+    tree_shapes = _tree_map_with_path(_standardize_shapes, combined_args, dynamic_shapes, tree_name="inputs")
+    serialized_shapes = tree_map(_track_dim_from_dims, tree_shapes)
+    return _postprocess_serialized_shapes(serialized_shapes, dims)
+
+
+def _loads(spec):
+    """
+    Utility function for dynamic shapes serialization.
+    Deserializes a dynamic_shapes spec from _loads() or _extracts() into a dynamic_shapes input to export().
+    """
+    import sympy
+
+    dynamic_shapes = spec["dynamic_shapes"]
+    dims = spec["dims"]
+    if dynamic_shapes is None:
+        return None
+
+    dim_cache = {}
+    for name, info in dims.items():
+        symbol = sympy.sympify(name)
+        dim_cache[name] = Dim(name, min=info["min"], max=info["max"])  # cache root dim
+        for _expr in info["derived"]:
+            expr = sympy.sympify(_expr)
+            modulus, remainder = sympy.polys.polytools.div(expr, symbol)
+            ddim = dim_cache[name]
+            if modulus != 1:
+                ddim = int(modulus) * ddim
+            if remainder != 0:
+                ddim = ddim + int(remainder)
+            dim_cache[_expr] = ddim  # cache derived dims
+
+    def deserialize_shape(val):
+        if val is None or isinstance(val, int):
+            return val
+        elif val == "_DimHint.AUTO":
+            return _DimHint.AUTO
+        elif val == "_DimHint.STATIC":
+            return _DimHint.STATIC
+        assert isinstance(val, str)
+        return dim_cache[val]
+
+    return tree_map(deserialize_shape, dynamic_shapes)
+
+
+def _extracts(ep: ExportedProgram):
```

Comment:
Would it be more composable to have a function which get the normal dynamic_shapes dictionary? And then users can call dump/load on that directly?

Reply:
Oh, users can call `_loads(_extracts(ep))`, I guess it's just the extra step

Reply:
I'd prefer `_extracts` to be in a separate PR, and probably as a method on `ExportedProgram` that has the behavior @angelayi suggested.

Reply:
Also (didn't read below yet), how do you keep the names `dx` etc. around, do you instead use symbols like `s0`?

Reply:
ah the names disappear, right now I just enumerate them into `d0, d1, ...` (using the symbols seem to have some strict/non-strict mismatches). I'll split this into a separate PR.